### PR TITLE
Add CPU and Memory modal

### PR DESCRIPTION
**Pull request by avivtur**
## 📝 Description

Add CPU and Memory edit modal

## 🎥 Demo

cpu-modal-vm-running

cpu-modal-vm-stopped

Signed-off-by: Aviv Turgeman <aturgema@redhat.com>

### CODE REVIEW

File: src/utils/components/CPUMemoryModal/CpuMemoryModal.tsx [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+import './cpu-memory-modal.scss';
+
+type CPUMemoryModalProps = {
+  vm: V1VirtualMachine;
+  isOpen: boolean;
+  onClose: () => void;
+};
+
+const CPUMemoryModal: React.FC<CPUMemoryModalProps> = ({ vm, isOpen, onClose }) => {
```
Comment (glekner) [inline src/utils/components/CPUMemoryModal/CpuMemoryModal.tsx:33]: we need to add an `onSubmit` prop

`onSubmit: (updatedVM: V1VirtualMachine) => Promise<void>`

File: src/utils/components/CPUMemoryModal/CpuMemoryModal.tsx [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+      title={t('Edit CPU | Memory')}
+      isOpen={isOpen}
+      className="cpu-memory-modal"
+      variant={ModalVariant.small}
+      onClose={onClose}
+      actions={[
+        <Button
+          key="confirm"
+          variant={ButtonVariant.primary}
+          onClick={onSubmit}
```
Comment (glekner) [inline src/utils/components/CPUMemoryModal/CpuMemoryModal.tsx:101]: `onClick={() => onSubmit(updatedVirtualMachine)}`

File: src/utils/components/CPUMemoryModal/CpuMemoryModal.tsx [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+  const [memory, setMemory] = React.useState<number>();
+  const [cpuCores, setCpuCores] = React.useState<number>();
+  const [memoryUnit, setMemoryUnit] = React.useState<string>();
+  const [isDropdownOpen, setIsDropdownOpen] = React.useState<boolean>(false);
+
+  const resultVirtualMachine = React.useMemo(() => {
+    const updatedVM = produce<V1VirtualMachine>(vm, (vmDraft: V1VirtualMachine) => {
+      vmDraft.spec.template.spec.domain.resources.requests = {
+        ...vm?.spec?.template?.spec?.domain?.resources?.requests,
```
Comment (glekner) [inline src/utils/components/CPUMemoryModal/CpuMemoryModal.tsx:51]: why do u need this spread? `immer` takes care of that.
I would also use `ensurePath` to make sure the path is existing.

```
const updatedVM = produce<V1VirtualMachine>(vm, (vmDraft: V1VirtualMachine) => {
    ensurePath(vmDraft, ['spec.template.spec.domain.resources.requests', 'spec.template.spec.domain.cpu'])

    vmDraft.spec.template.spec.domain.resources.requests.memory = `${memory}${memoryUnit}`;
    vmDraft.spec.template.spec.domain.cpu.cores = cpuCores;
});

File: src/utils/components/CPUMemoryModal/CpuMemoryModal.tsx [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        [key: string]: string;
+      };
+      const { size, unit } = getMemorySize(requests?.memory);
+      setMemoryUnit(unit);
+      setMemory(size);
+      setCpuCores(getCPUCount(vm?.spec?.template?.spec?.domain?.cpu));
+    }
+  }, [vm]);
+
+  const onSubmit = () => {
```
Comment (glekner) [inline src/utils/components/CPUMemoryModal/CpuMemoryModal.tsx:71]: delete this

File: src/utils/components/CPUMemoryModal/CpuMemoryModal.tsx [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+  } = useTemplateDefaultCpuMemory(vm);
+  const [updateInProcess, setUpdateInProcess] = React.useState<boolean>(false);
+  const [updateError, setUpdateError] = React.useState<string>();
+
+  const [memory, setMemory] = React.useState<number>();
+  const [cpuCores, setCpuCores] = React.useState<number>();
+  const [memoryUnit, setMemoryUnit] = React.useState<string>();
+  const [isDropdownOpen, setIsDropdownOpen] = React.useState<boolean>(false);
+
+  const resultVirtualMachine = React.useMemo(() => {
```
Comment (glekner) [inline src/utils/components/CPUMemoryModal/CpuMemoryModal.tsx:48]: `const updatedVirtualMachine`

File: src/utils/components/CPUMemoryModal/CpuMemoryModal.tsx (hunk 3 above)
Comment (avivtur) [reply src/utils/components/CPUMemoryModal/CpuMemoryModal.tsx:51]: `resources` is of type object or to be exact: `{}` like `any` so `requests.memory` is not recognized as a type,
had to do it to keep previous values if there are any

File: src/utils/components/CPUMemoryModal/CpuMemoryModal.tsx (hunk 4 above)
Comment (avivtur) [reply src/utils/components/CPUMemoryModal/CpuMemoryModal.tsx:71]: error and loading states are being managed by the onSubmit, should I get those as props instead?

**Pull request merged**